### ISSUE TITLE
Pytree structure of distributions with `lazy_property` depends on previous attribute access.

### ISSUE THREAD

**Issue by tillahoffmann** [pyro-ppl/numpyro#1928]
The pytree structure of distributions with `lazy_property` depends on which attributes have previously been accessed, which can cause trouble with `jax.lax.scan`. This is probably most easily explained with an example.

```python
>>> import jax
>>> from jax import numpy as jnp
>>> from numpyro import distributions as dists
>>> 
>>> 
>>> def make_dist():
>>>     return dists.MultivariateNormal(precision_matrix=jnp.eye(2))
>>> 
>>> 
>>> init = make_dist()
>>> init.covariance_matrix  # No error if this line is commented.
>>> jax.lax.scan(lambda carry, _: (make_dist(), None), init, jnp.arange(7))
---------------------------------------------------------------------------
TypeError: Scanned function carry input and carry output must have the same 
pytree structure, but they differ:
  * the input carry component carry[<flat index 1>] is a 
    <class 'jaxlib.xla_extension.ArrayImpl'> but the corresponding component 
    of the carry output is a <class 'NoneType'>, so their Python types differ

Revise the scanned function so that its output is a pair where the first 
element has the same pytree structure as the first argument.
```

Before accessing `covariance_matrix`, the field does not appear in `__dict__` such that the flattening code replaces the field by `None` here.

https://github.com/pyro-ppl/numpyro/blob/244ff38c727633f8ab5a67ee24444461fc6d3954/numpyro/distributions/distribution.py#L173-L179

Upon accessing `covariance_matrix`, the `lazy_property` is replaced by a value which in turn replaces the `None` in the pytree representation.

```python
>>> d = make_dist()
>>> "covariance_matrix" in d.__dict__
False
>>> d.covariance_matrix
Array([[1., 0.],
       [0., 1.]], dtype=float32)
>>> "covariance_matrix" in d.__dict__
True
```

For each distribution, it seems there is a set of canonical parameters which are not `lazy_property`s (`scale_tril` in the case of the `MultivariateNormal` distribution). Would it make sense to only include names in `arg_constraints` if the corresponding element on the distribution class is not a `lazy_property`? I'm thinking of these lines:

https://github.com/pyro-ppl/numpyro/blob/244ff38c727633f8ab5a67ee24444461fc6d3954/numpyro/distributions/distribution.py#L146-L159

Happy to make the change if that seems like a reasonable approach. A downside might be that caching of `lazy_property`s does not persist across jitted functions.

**Comment by fehiepsi** [comment]
That's a nice observation. I think the approach makes sense. Could you make a PR to stimulate the discussion?

**Comment by tillahoffmann** [comment]
Fixed in #1929.
